### PR TITLE
[2.0] Refactor the error and exception handlers

### PR DESCRIPTION
**Pull request by ste93cry**
| Q             | A
| ------------- | ---
| Branch?       | 2.0
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

This PR is a huge refactoring of the error and exception handlers. I tried to make them as transparent as possible in regards to other handlers, so the error handler converts the errors to exceptions to be able to log them but never throws them unless necessary. The aim of my work is to fix the many bugs that we're experiencing in `1.x`, in particular #579, #552, #391, #343. I'm currently tracking an issue with Zend Diactoros (zendframework/zend-diactoros#298) that prevents any error past the first one from being logged through custom error handlers after usage of the `Stream` class.

### CODE REVIEW

File: lib/Raven/ErrorHandler.php [hunk 1]
```diff
@@ -11,203 +9,295 @@
...
-    }
 
-    public function shouldCaptureFatalError($type)
-    {
-        // Do not capture E_ERROR since those can be caught by userland since PHP 7.0
-        // E_ERROR should already be handled by the exception handler
-        // This prevents duplicated exceptions in PHP 7.0+
-        if (PHP_VERSION_ID >= 70000 && E_ERROR === $type) {
-            return false;
+        if (!empty($error) && $error['type'] & (E_ERROR | E_PARSE | E_CORE_ERROR | E_CORE_WARNING | E_COMPILE_ERROR | E_COMPILE_WARNING)) {
```
Comment (Jean85) [inline lib/Raven/ErrorHandler.php]: Maybe you should wrap those two lines with a private method to give it a name?

File: tests/ErrorHandlerTest.php [hunk 2]
```diff
@@ -12,239 +12,311 @@
...
     }
 
-    public function tearDown()
+    /**
+     * @dataProvider constructorThrowsWhenReservedMemorySizeIsWrongDataProvider
+     *
+     * @expectedException \UnexpectedValueException
+     * @expectedExceptionMessage The value of the $reservedMemorySize argument must be an integer greater than 0.
+     */
+    public function testConstructorThrowsWhenReservedMemorySizeIsWrong()
```
Comment (Jean85) [inline tests/ErrorHandlerTest.php]: You're not using the data from the provider

File: lib/Raven/ErrorHandler.php (hunk 1 above)
Comment (ste93cry) [reply lib/Raven/ErrorHandler.php]: Extracting the code that just checks whether the error is a fatal error does not makes much sense imho, and there would also be the cost of calling one more function

File: lib/Raven/ErrorHandler.php [hunk 3]
```diff
@@ -11,203 +9,331 @@
...
+    public function handleException($exception)
     {
-        $this->old_exception_handler = set_exception_handler([$this, 'handleException']);
-        $this->call_existing_exception_handler = $call_existing;
+        $this->client->captureException($exception);
+
+        // If there is no exception handler that can run after the current one
+        // give back the exception to the native PHP handler
+        if (null === $this->previousExceptionHandler) {
+            throw $exception;
```
Comment (Jean85) [inline lib/Raven/ErrorHandler.php]: Doesn't re-throwing change the stack trace of the exception?

File: lib/Raven/ErrorHandler.php [hunk 4]
```diff
@@ -11,203 +9,331 @@
...
+
+        $previousExceptionHandlerException = $exception;
+
+        // Unset the previous exception handler to prevent infinite loop in case
+        // we need to handle an exception thrown from it
+        $previousExceptionHandler = $this->previousExceptionHandler;
+        $this->previousExceptionHandler = null;
+
+        try {
+            call_user_func($previousExceptionHandler, $exception);
```
Comment (Jean85) [inline lib/Raven/ErrorHandler.php]: `$previousExceptionHandler($exception)` is maybe slightly faster?

File: lib/Raven/ErrorHandler.php [hunk 5]
```diff
@@ -11,203 +9,331 @@
...
+
+        // If the exception object instance is the same as the one catched from
+        // the previous exception handler, if any, give it back to the native
+        // PHP handler to prevent infinite circular loop
+        if ($exception === $previousExceptionHandlerException) {
+            throw $exception;
+        }
 
-        return $this;
+        $this->handleException($previousExceptionHandlerException);
```
Comment (Jean85) [inline lib/Raven/ErrorHandler.php]: Is this loop-prone?

File: tests/phpt/fatal_error.phpt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+ErrorHandler::register($client);
+
+register_shutdown_function('register_shutdown_function', function () use ($client) {
+    /** @var \Raven\Transport\HttpTransport $transport */
+    $transport = Assert::getObjectAttribute($client, 'transport');
+    $pendingRequests = Assert::getObjectAttribute($transport, 'pendingRequests');
+
+    Assert::assertEmpty($pendingRequests);
+});
```
Comment (Jean85) [inline tests/phpt/fatal_error.phpt]: Those asserts can trigger the autoloading. If this is done under PHP 5 inside an error handling context, it fails silently. You should add an `echo` at the end and expect it to be sure that all those assertions are executed correctly.

File: tests/phpt/out_of_memory.phpt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+require $vendor . '/vendor/autoload.php';
+
+$client = ClientBuilder::create()->getClient();
+
+ErrorHandler::register($client);
+
+$foo = str_repeat('x', 1024 * 1024 * 30);
+?>
+--EXPECTF--
+Fatal error: Allowed memory size of %d bytes exhausted (tried to allocate %d bytes) in %s on line %d
```
Comment (Jean85) [inline tests/phpt/out_of_memory.phpt]: Shouldn't you add something to be sure that you really handled the error?

File: lib/Raven/ErrorHandler.php (hunk 5 above)
Comment (ste93cry) [reply lib/Raven/ErrorHandler.php]: Yes it should be as the next time the function will be called `$this->previousExceptionHandler` will be `null` and after sending the exception to Sentry the exception itself will be rethrown

File: lib/Raven/ErrorHandler.php (hunk 3 above)
Comment (ste93cry) [reply lib/Raven/ErrorHandler.php]: After a few tests it seems it doesn't

File: tests/phpt/out_of_memory.phpt (hunk 7 above)
Comment (ste93cry) [reply tests/phpt/out_of_memory.phpt]: Done, I've added an `echo` statement after all the assertions whose existence is checked in the `--EXPECTF--` section of the PHPT test

**Comment by ste93cry** [comment]
I still have to complete the refactor of the ErrorHandler for the breadcrumbs. Please don't merge this PR yet until I remove the WIP prefix

**Comment by Jean85** [comment]
Sorry, I thought it was complete!

**Comment by ste93cry** [comment]
>so an end user would still be able to customize the breadcrumbs.

Would it makes sense for an user to use only the error handler that registers the breadcrumbs and not the handler that captures the error itself?

**Comment by Jean85** [comment]
No, obviously not, why?

**Comment by ste93cry** [comment]
Because then it would not make sense to have two error handlers, one for capturing the errors and one for registering the breadcrumb. Or at least, I don't see the sense behind this. Having two handlers means  reserving memory for both of them (to handle fatal errors) and duplicating the logic for each of them. Also, I don't see why an user would want to not register the breadcrumbs but instead capture the errors

**Comment by Jean85** [comment]
The scenario that I want to preserve is the possibility of replacing (or enhancing) the breadcrumb handler. If you collapse both handler this would not be possible anymore.

**Comment by ste93cry** [comment]
I updated the PR by refactoring the breadcrumb handler and it's ready for another round of code reviews. The [zendframework/zend-diactoros#298](https://github.com/zendframework/zend-diactoros/issues/298) issue persists and if nothing changes before releasing `2.0` we will have to switch PSR-7 library or no errors after the first one will be logged and sent to the Sentry server

**Comment by ste93cry** [comment]
I fixed the fact that fatal errors were reported twice in PHP 7+ (yes, again...) and this should be the good time. Before merging, should I switch PSR-7 library in this PR or do you prefer to postpone it later?

**Pull request merged**